pkg/controllers/clusters/clusterlifecycle/clusterlifecycle_controller.go

```diff
@@ -0,0 +1,354 @@
+	if mcls.Status.HeartbeatFrequencySeconds != nil {
+		requeueAfter = time.Second*time.Duration(*mcls.Status.HeartbeatFrequencySeconds)*ClusterNotReadyThreshold + 10*time.Second
+	}
+	return requeueAfter, nil
+}
+
+// enqueue takes a ManagedCluster resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than ManagedCluster.
+func (c *Controller) enqueue(cluster *clusterapi.ManagedCluster) {
+	key, err := cache.MetaNamespaceKeyFunc(cluster)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+func (c *Controller) updateClusterToLostIfNeeded(ctx context.Context, cluster *clusterapi.ManagedCluster) error {
+	currentReadyCondition := GetClusterCondition(&cluster.Status, clusterapi.ClusterReady)
+	observedReadyCondition := c.getClusterReadyCondition(cluster)
+
+	if observedReadyCondition != nil && !equality.Semantic.DeepEqual(currentReadyCondition, observedReadyCondition) {
+		// update Readyz,Livez,Healthz to false when cluster status is unknown
+		if observedReadyCondition.Status == metav1.ConditionUnknown {
+			cluster.Status.Readyz = false
+			cluster.Status.Livez = false
+			cluster.Status.Healthz = false
+		}
+
+		cluster.Status.Conditions = []metav1.Condition{*observedReadyCondition}
+
+		_, err := c.clusternetClient.ClustersV1beta1().ManagedClusters(cluster.Namespace).UpdateStatus(ctx, cluster, metav1.UpdateOptions{})
+		if err != nil {
+			msg := fmt.Sprintf("failed to update conditions of ManagedCluster status: %v", err)
+			klog.WarningDepth(2, msg)
+			c.recorder.Event(cluster, corev1.EventTypeWarning, "FailedUpdatingClusterStatus", msg)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getClusterReadyCondition return cluster unknown condition based on cluster status and grace period
+func (c *Controller) getClusterReadyCondition(cluster *clusterapi.ManagedCluster) *metav1.Condition {
+	var readyCondition *metav1.Condition
+	var probeTimestamp metav1.Time
+
+	gracePeriod := DefaultClusterMonitorGracePeriod
+	if cluster.Status.HeartbeatFrequencySeconds != nil {
+		gracePeriod = time.Second * time.Duration(*cluster.Status.HeartbeatFrequencySeconds) * ClusterNotReadyThreshold
+	}
+
+	lastObservedTime := cluster.Status.LastObservedTime
+	if lastObservedTime.IsZero() {
+		// If lastObservedTime is zero means Clusternet agent never posted cluster status.
+		// We treat this cluster to be a new cluster and use cluster.CreationTimestamp to be probeTimestamp.
+		probeTimestamp = cluster.CreationTimestamp
+	} else {
+		probeTimestamp = lastObservedTime
+	}
+
+	if metav1.Now().After(probeTimestamp.Add(gracePeriod)) {
+		if lastObservedTime.IsZero() {
+			// Clusternet agent never posted cluster status and reach grace period
+			return &metav1.Condition{
+				Type:               clusterapi.ClusterReady,
+				Status:             metav1.ConditionUnknown,
```

Comment:
The `Reason` and `Status` in condition are quite misleading.

Normally `Reason` and `Message` have the same meanings, and do not get changed.